docs/source/migration.mdx

```diff
@@ -573,24 +604,47 @@ const server = new ApolloServer({
 ```
 
 ### `formatResponse` hook
-In Apollo Server 3, the `formatResponse` hook enables you to transform the structure of GraphQL response objects before they're sent to a client.
 
-Apollo Server 4 removes the `formatResponse` hook. We instead recommend using the `willSendResponse` plugin to do the same thing:
+In Apollo Server 3, the `formatResponse` hook enables you to transform the structure of GraphQL response objects before they're sent to a client. This hook is called only after an operation executes (not if there is an error like a parse error that prevents execution). It receives the `response` `GraphQLResponse` and the `requestContext` (which does not yet contain the `response`). If it returns a non-null `GraphQLResponse`, that response is used instead of the one it received. This hook is provided as a top-level constructor argument, not as part of the plugin API.
+
+There is also a `willSendResponse` plugin hook. This plugin hook is called at a similar place in the lifecycle to `formatResponse`. It receives a `requestContext`; the `GraphQLResponse` is on this object in the `response` field. It may mutate `requestContext.response`. It is called for all requests that get far enough to invoke `requestDidStart` (ie, any request where the JSON body can be properly parsed, etc). Anything that the `formatResponse` hook can do can also be accomplished with the `willSendResponse` plugin hook.
 
-TODO: write as before and after
+Apollo Server 4 removes the `formatResponse` hook. We instead recommend using the `willSendResponse` plugin to do the same thing. (Note that [the structure of `GraphQLResponse` has also changed](#graphqlresponse) so that most of its fields are nested under a `result` field.) You can tell if an operation made it to the execution phase by seeing if the result has a `data` field, so if it was important that your `formatResponse` hook is only called if execution occurred, you can check for the existence of this field. (There are some small edge cases here: for example, an error in the coercion of variable values would result in calling `formatResponse` but will not have `data` in the result. If differentiating these edge cases is important to you, open an issue and we'll help.)
```

Comment:
Take the information from the old paragraph above this one and combining them together:

```suggestion
Apollo Server 4 removes the `formatResponse` hook. We instead recommend using the `willSendResponse` plugin hook, which enables you to do everything you previously did with `formatResponse`. Apollo Server calls the  `willSendResponse` plugin hook for all requests that get far enough to invoke `requestDidStart` (i.e., requests with a parsable JSON body, etc. ). 

The `willSendResponse` plugin hook receives an operation's `requestContext`, which has a `response` field containing a `GraphQLResponse` object. Note that the `willSendResponse` plugin hook might mutate the `requestContext.response` field. 

> Note that Apollo Server 4 changes the structure of `GraphQLResponse`; [see below for more details](#graphqlresponse). 
```

Is this extra background on `formatResponse` necessary:
> You can tell if an operation made it to the execution phase by seeing if the result has a `data` field, so if it was important that your `formatResponse` hook is only called if execution occurred, you can check for the existence of this field. (There are some small edge cases here: for example, an error in the coercion of variable values would result in calling `formatResponse` but will not have `data` in the result. If differentiating these edge cases is important to you, open an issue and we'll help.)

It doesn't seem to apply to `willSendResponse`, and  if folks are no longer using `formatResponse`, then it feels like maybe this section could be excluded?

Reply:
willSendResponse gets called in more contexts than formatResponse, so if you care about maintaining the old behavior you have to implement the filter yourself.

You changed "may" to "might", but I meant "may" in the sense of "is allowed to".

Reply:
Thank for your the clarification, I've updated the below suggestion to incorporate the intentional "may", and the extra info on `willSendResponse` being called more: 

```suggestion
Apollo Server 4 removes the `formatResponse` hook. We instead recommend using the `willSendResponse` plugin hook, which enables you to do everything you previously did with `formatResponse`. The `willSendResponse` plugin hook receives an operation's `requestContext`, which has a `response` field containing the `GraphQLResponse` object. Note that the `willSendResponse` hook may mutate the `requestContext.response` field. 

> Note that Apollo Server 4 changes the structure of `GraphQLResponse`; [see below for more details](#graphqlresponse). 

Apollo Server calls the `willSendResponse` plugin hook for all requests that get far enough to invoke `requestDidStart` (i.e., requests with a parsable JSON body, etc. ). This means that Apollo Server calls the `willSendResponse` hook in more contexts than the previous `formatResponse` hook. 

To only use the `willSendResponse` hook after an operation's "execution" stage (i.e., like the previous `formatResponse` hook), you can make a filter checking for the existence of a `data` field in the result. If an operation has a `data` field in the result, it has made it to the execution phase. Note, there are some edge cases: for example, an error in the coercion of variable values calls `formatResponse` but doesn't have `data` in the result. If differentiating these edge cases is important to you, please open an issue, and we'll help. 
```